devito/ir/support/utils.py

```diff
@@ -0,0 +1,193 @@
+
+
+def detect_flow_directions(exprs):
+    """Return a mapper from :class:`Dimension`s to iterables of
+    :class:`IterationDirection`s representing the theoretically necessary
+    directions to evaluate ``exprs`` so that the information "naturally
+    flows" from an iteration to another."""
+    exprs = as_tuple(exprs)
+
+    writes = [Access(i.lhs, 'W') for i in exprs]
+    reads = flatten(retrieve_indexed(i.rhs, mode='all') for i in exprs)
+    reads = [Access(i, 'R') for i in reads]
+
+    # Determine indexed-wise direction by looking at the vector distance
+    mapper = defaultdict(set)
+    for w in writes:
+        for r in reads:
+            if r.name != w.name:
+                continue
+            dimensions = [d for d in w.aindices if d is not None]
+            for d in dimensions:
+                try:
+                    if w.distance(r, d) > 0:
+                        mapper[d].add(Forward)
+                        break
+                    elif w.distance(r, d) < 0:
+                        mapper[d].add(Backward)
+                        break
+                    else:
+                        mapper[d].add(Any)
+                except TypeError:
+                    # Nothing can be deduced
+                    mapper[d].add(Any)
+                    break
+            # Remainder
+            for d in dimensions[dimensions.index(d) + 1:]:
+                mapper[d].add(Any)
+
+    # Add in any encountered Dimension
+    mapper.update({d: {Any} for d in flatten(i.aindices for i in reads + writes)
+                   if d is not None and d not in mapper})
+
+    # Add in stepping dimensions (just in case they haven't been detected yet)
+    # note: stepping dimensions may force a direction on the parent
+    assert all(v == {Any} or mapper.get(k.parent, v) in [v, {Any}]
+               for k, v in mapper.items() if k.is_Stepping)
+    mapper.update({k.parent: set(v) for k, v in mapper.items()
+                   if k.is_Stepping and mapper.get(k.parent) == {Any}})
+
+    # Add in derived dimensions parents
+    mapper.update({k.parent: set(v) for k, v in mapper.items()
+                   if k.is_Derived and k.parent not in mapper})
+
+    return mapper
+
+
+def compute_directions(exprs, key):
+    """
+    Return a mapper ``M : D -> I`` where D is the set of :class:`Dimension`s
+    found in the input expressions ``exprs``, while I = {Any, Backward,
+    Forward} (i.e., the set of possible :class:`IterationDirection`s).
+
+    The iteration direction is chosen so that the information "naturally flows"
+    from an iteration to another (i.e., to generate "flow" or "read-after-write"
+    dependencies).
+
+    In the case of a clash (e.g., both Forward and Backward should be used
+    for a given dimension in order to have a flow dependence), the function
+    ``key : D -> I`` is used to pick one value.
+    """
+    mapper = detect_flow_directions(exprs)
+    return force_directions(mapper, key)
```

Comment:
I am not sure why we need this function, seems like `force_directions`  alone is enough

Reply:
``detect`` doesn't make a choice; ``force`` makes a choice. ``detect`` may be computationally expensive, so it's handy to keep them separated (see ``clusters/algorithms.py``)